### PR TITLE
Fixed to reading/writing of external records

### LINKED ISSUE #642: Is the external type validate part correct?
image

asciiDomain here is the NSS part of https://datatracker.ietf.org/doc/html/rfc2141.html#section-2.2

For instance there is the rule that if there is a % it must be followed by two hex numbers, that is not checked in the algorithm.

Apart from that  "/" | "?" | "#" are allowed but reserved so we should probably ignore those for now (which we are doing).

### PR DESCRIPTION
**Pull request by kenchris**
Closes #642

***
<a href="https://pr-preview.s3.amazonaws.com/kenchris/web-nfc/pull/643.html" title="Last updated on Jun 29, 2022, 10:58 AM UTC (187e2d5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/web-nfc/643/78f8fb0...kenchris:187e2d5.html" title="Last updated on Jun 29, 2022, 10:58 AM UTC (187e2d5)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -2828,46 +2828,43 @@ <h3><dfn>Writing content</dfn></h3>
...
-          <a href="https://url.spec.whatwg.org/#concept-domain-to-ascii">
-          domain to ASCII</a> given |domain| and `true` (as |beStrict|).
+          Return the |domain| and |type| pair.
         </li>
+      </ol>
+      <p>
+        To <dfn>validate external type</dfn> given |input:USVString|, run
+        these steps:
+      </p>
+      <ol class=algorithm id="split-external-type">
```
Comment (beaufortfrancois) [inline index.html]: I believe `id` should be `"validate-external-type"`

File: index.html [hunk 2]
```diff
@@ -2828,46 +2828,43 @@ <h3><dfn>Writing content</dfn></h3>
...
-        To <dfn>validate external type</dfn> given |input:USVString|, run
+        To <dfn>split external type</dfn> given |input:USVString|, run
         these steps:
       </p>
-      <ol class=algorithm id="validate-external-type">
-        <li>
-          If |input:string| is not an [=ASCII string=], is empty, or its
-          length exceeds 255 bytes, return `false`.
-        </li>
+      <ol class=algorithm id="split-external-type">
```
Comment (beaufortfrancois) [inline index.html:2834]: Shall we fail early when `:` is not found with something like this?
> 1.  If `U+003A` (`:`) is not found in |input|, return failure.
> 2. Let |domain| be the |input| from the start of |input| up to but excluding the first occurrence of `U+003A` (`:`).
> 3. Let |type| be the |input| after the first occurrence of `U+003A` (`:`) up to the end of |input|.
> 4. Return the |domain| and |type| pair.

File: index.html [hunk 3]
```diff
@@ -2828,46 +2828,43 @@ <h3><dfn>Writing content</dfn></h3>
...
         <li>
-          If |asciiDomain| is failure, return `false`.
+          Let |domain| and |type| be the result of running [=split external type=], or
+          return `false` if failure.
         </li>
         <li>
-          If |asciiDomain| contains a [=forbidden host code point=] or
-          `U+005F` `LOW LINE` (`_`), return `false`.
+          If |domain| is not a [=valid domain string=], return failure.
```
Comment (beaufortfrancois) [inline index.html]: Why not returning `false` as before in the case of failure?

File: index.html [hunk 4]
```diff
@@ -2828,46 +2828,43 @@ <h3><dfn>Writing content</dfn></h3>
...
         </li>
         <li>
           If |type| contains [=code points=] that are not
           [=ASCII alphanumeric=], or `U+0024` (`$`), `U+0027` (`'`),
           `U+0028` `LEFT PARENTHESIS` (`(`), `U+0029` `RIGHT PARENTHESIS` (`)`),
           `U+002A` (`*`), `U+002B` (`+`), `U+002C` (`,`), `U+002D` (`-`),
           `U+002E` (`.`), `U+003B` (`;`), `U+003D` (`=`), `U+0040` (`@`),
-          `U+005F` (`_`), return `false`.
+          `U+005F` (`_`), return failure.
```
Comment (beaufortfrancois) [inline index.html]: Why not returning `false` as before in the case of failure?

File: index.html [hunk 5]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
...
+          <li>
+            If |customTypeName| is not a {{USVString}} or its length exceeds 255 bytes, [= exception/throw =] a {{TypeError}}.
+          </li>
+          <li>
+            Set the |ndef|'s <a>TYPE field</a> to |customTypeName|.
           </li>
           <li>
             If the type of a |record|'s <a>data</a> is not
             {{BufferSource}} or {{NDEFMessageInit}}, [= exception/throw =] a
-            {{TypeError}} and abort these steps.
```
Comment (beaufortfrancois) [inline index.html]: I've noticed we're not consistent the spec when it comes to "throw a TypeError" and "throw a TypeError and abort these steps". Please file a separate issue to address those and not pollute this PR.

File: index.html [hunk 6]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
         To <dfn>map external data to NDEF</dfn> given |record:NDEFRecordInit|,
         |ndef|, and |recordsDepth:unsigned short|, run these steps:
         <ol class=algorithm data-link-for="NDEFRecordInit">
-          <!--
```
Comment (beaufortfrancois) [inline index.html]: Why did you remove this comment? Is this intentional?

File: index.html [hunk 7]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
...
           <li>
-            If |context| is `"smart-poster"`,
-            [= exception/throw =] a {{TypeError}} and abort these steps.
+            If |record|'s <a>mediaType</a> is not `undefined`,
+            [= exception/throw =] a {{TypeError}}.
```
Comment (beaufortfrancois) [inline index.html]: Same here. Let's not remove the "and abort these steps" in this PR.

File: index.html [hunk 8]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
...
-            [= exception/throw =] a {{TypeError}} and abort these steps.
+            Let |domain| and |type| be the result of running [=split external type=] with
+            |record|'s <a>recordType</a>.
+          </li>
+          <li>
+            Let |domain| be the result of running <a href="https://url.spec.whatwg.org/#concept-domain-to-ascii">
+            domain to ASCII</a> with |domain| and true.
+          </li>
+          <aside class="note">
+            The recordType is user set and in UTF-8 encoding, allowing for international domain names as part of the
```
Comment (beaufortfrancois) [inline index.html]: Nit: `"The record's <a>recordType</a> is..."`

File: index.html [hunk 9]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
...
+            custom type such as "håndværker.dk:abc". When writing that to the NDEF it has to be converted to ASCII
```
Comment (beaufortfrancois) [inline index.html]: Nit: `"When writing that value to an NDEF record"`

File: index.html [hunk 10]
```diff
@@ -3243,29 +3240,41 @@ <h3><dfn>Writing content</dfn></h3>
...
+            bytes and after the conversion this would be represented as "xn--hndvrker-9zan.dk:abc".
+          </aside>
+          <li>
+            Let |customTypeName| be |domain|, ":" and |type| concatenated.
```
Comment (beaufortfrancois) [inline index.html]: Nit: Use backticks for colon.

File: index.html [hunk 11]
```diff
@@ -4444,8 +4453,19 @@ <h3>Parsing content</h3>
...
+          Let |domain| be the result of running <a href="https://www.unicode.org/reports/tr46/#ToUnicode">Unicode ToUnicode</a>
+          with |domain_name| set to |domain|, |CheckHyphens| set to false, |CheckBidi| set to true, |CheckJoiners| set to true,
+          |UseSTD3ASCIIRules| set to true, and |Transitional_Processing| set to false.
+        </li>
+        <aside class="note">
+          The <a>TYPE field</a> contains a domain part stored as ASCII, so for instance "xn--hndvrker-9zan.dk:abc" would
+          be presented as "håndværker.dk:abc" after the above conversion.
+        </aside>
+        <li>
+          Set |record|'s <a>recordType</a> to |domain|, ":" and |type| concatenated.
```
Comment (beaufortfrancois) [inline index.html]: Nit: Use backticks for colon.

File: index.html [hunk 12]
```diff
@@ -4444,8 +4453,19 @@ <h3>Parsing content</h3>
       a |record:NDEFRecord|, run these steps:
       <ol class=algorithm data-link-for="NDEFRecord">
         <li>
-          Set |record|'s <a>recordType</a> to the
-          value of |ndefRecord|'s <a>TYPE field</a>.
+          Let |domain| and |type| be the result of running [=split external type=] with the value of |ndefRecord|'s <a>TYPE field</a>.
```
Comment (beaufortfrancois) [inline index.html:4469]: We should return null in cases of failure so that the record is not added to message's records.

**Review by beaufortfrancois — CHANGES_REQUESTED** [review]
Thanks for the PR @kenchris!

File: index.html (hunk 2 above)
Comment (kenchris) [reply index.html:2834]: Yeah makes sense

File: index.html (hunk 3 above)
Comment (kenchris) [reply index.html]: My mistake :-) I guess I should have paid more attention

File: index.html (hunk 6 above)
Comment (kenchris) [reply index.html]: Yes, I really don't see what smart-poster has to do with this, I guess it was copied earlier to use the same language

File: index.html (hunk 7 above)
Comment (kenchris) [reply index.html]: OK, you don't need to say that when throwing though, but I guess we could fix separately... same as `false` should just be false etc

File: index.html (hunk 12 above)
Comment (kenchris) [reply index.html:4469]: In this case, they have been validated earlier so they will always be valid

File: index.html (hunk 2 above)
Comment (zolkis) [reply index.html:2834]: I don't know. To me it sounds more intuitive to have step 2 first since that is the real operation, and then step 1 as an error step of the former, otherwise someone might think they have to scan the string separately to check for `:`? 
No strong opinion here.

File: index.html [hunk 13]
```diff
@@ -3243,15 +3241,31 @@ <h3><dfn>Writing content</dfn></h3>
...
-            [= exception/throw =] a {{TypeError}} and abort these steps.
+            Let |domain| and |type| be the result of running [=split external type=] with
+            |record|'s <a>recordType</a>.
+          </li>
+          <li>
+            Let |domain| be the result of running <a href="https://url.spec.whatwg.org/#concept-domain-to-ascii">
+            domain to ASCII</a> with |domain| and true.
+          </li>
+          <aside class="note">
+            The record's <a>recordType</a> is is user set and in UTF-8 encoding, allowing for international domain names as part of the
```
Comment (zolkis) [inline index.html]: typo: s/is is/is

File: index.html [hunk 14]
```diff
@@ -4444,8 +4454,22 @@ <h3>Parsing content</h3>
       a |record:NDEFRecord|, run these steps:
       <ol class=algorithm data-link-for="NDEFRecord">
         <li>
-          Set |record|'s <a>recordType</a> to the
-          value of |ndefRecord|'s <a>TYPE field</a>.
+          Let |domain| and |type| be the result of running [=split external type=] with the value of |ndefRecord|'s <a>TYPE field</a>.
+        </li>
+        <aside class="note">
+          The above step is ensures to work due to earlier validation of the external type.
```
Comment (zolkis) [inline index.html]: typo: "is ensures"?

File: index.html (hunk 13 above)
Comment (kenchris) [inline index.html]:
```suggestion
            The record's <a>recordType</a> is user set and in UTF-8 encoding, allowing for international domain names as part of the
```

**Review by zolkis — APPROVED** [review]
Nice, thanks.

File: index.html (hunk 2 above)
Comment (kenchris) [reply index.html:2834]: People can implement it that way as the result is the same, this is about clear spec text :-)

File: index.html (hunk 6 above)
Comment (beaufortfrancois) [reply index.html]: @zolkis added that comment as part of https://github.com/w3c/web-nfc/commit/24f5f88a074ecde6c6eccada12dff434de031ce8

File: index.html (hunk 12 above)
Comment (beaufortfrancois) [reply index.html:4469]: Where have they been validated earlier?

**Review by beaufortfrancois — CHANGES_REQUESTED** [review]

File: index.html (hunk 6 above)
Comment (zolkis) [reply index.html]: IIRC the algorithm was reused in multiple contexts. When the context was "smart-poster", the algorithm was in the course of building a smart poster record. So it's the mode of the algorithm in the context it was called.

If that is confusing, we might want to revisit that and normalize by removing context and repeating the algorithms in the spec. I think I attempted that and didn't like the result, but we could give it a go.

**Comment by kenchris** [comment]
> Where have they been validated earlier?

Actually it hasn't that was the other way around :/ So le me move the comment there and fix

File: index.html (hunk 6 above)
Comment (beaufortfrancois) [reply index.html]: Let's not remove this as part of this PR and file an issue to clean this if necessary.

File: index.html [hunk 15]
```diff
@@ -3243,15 +3241,34 @@ <h3><dfn>Writing content</dfn></h3>
...
           </li>
-          -->
           <li>
-            If |record|'s <a>mediaType</a> is not `undefined`,
-            [= exception/throw =] a {{TypeError}} and abort these steps.
+            Let |domain| and |type| be the result of running [=split external type=] with
+            |record|'s <a>recordType</a>.
+          </li>
+          <aside class="note">
+            The above step is ensures to work due to earlier validation of the external type.
```
Comment (beaufortfrancois) [inline index.html]: Nit: "The above step is ensured to..." or "The above step always succeeds due to..."

**Review by beaufortfrancois — CHANGES_REQUESTED** [review]

File: index.html [hunk 16]
```diff
@@ -3243,15 +3241,34 @@ <h3><dfn>Writing content</dfn></h3>
         To <dfn>map external data to NDEF</dfn> given |record:NDEFRecordInit|,
         |ndef|, and |recordsDepth:unsigned short|, run these steps:
         <ol class=algorithm data-link-for="NDEFRecordInit">
```
Comment (kenchris) [inline index.html]:
```suggestion
        <ol class=algorithm data-link-for="NDEFRecordInit">
          <!--
          <li>
            If |context| is `"smart-poster"`,
            If |record|'s <a>mediaType</a> is not `undefined`,
            [= exception/throw =] a {{TypeError}} and abort these steps.
          </li>
          -->
```

File: index.html [hunk 17]
```diff
@@ -3246,13 +3244,39 @@ <h3><dfn>Writing content</dfn></h3>
           <!--
           <li>
             If |context| is `"smart-poster"`,
+            If |record|'s <a>mediaType</a> is not `undefined`,
```
Comment (kenchris) [inline index.html]:
```suggestion
```

File: index.html [hunk 18]
```diff
@@ -3253,6 +3251,31 @@ <h3><dfn>Writing content</dfn></h3>
             If |record|'s <a>mediaType</a> is not `undefined`,
             [= exception/throw =] a {{TypeError}} and abort these steps.
           </li>
+          <li>
+            Let |domain| and |type| be the result of running [=split external type=] with
+            |record|'s <a>recordType</a>.
+          </li>
+          <aside class="note">
+            The above step is ensures to work due to earlier validation of the external type.
```
Comment (kenchris) [inline index.html]:
```suggestion
            The above step always succeeds due to earlier validation of the external type.
```

**Review by beaufortfrancois — APPROVED** [review]
Thank you @kenchris!
